### PR TITLE
Adjust the scream-scorpio interface freeing pio decomp memory when not needed

### PR DESCRIPTION
**Pull request by AaronDonahue**
This commit adds the new subroutine in the scream_scorpio_interface.F90 code,
`free_decomp`, which will free the memory in scorpio for any pio decomposition
that is no longer needed.  This routine is called whenever a file is closed.

There is also some updates to the infrastructure in the scream scorpio interface
to keep track of which PIO decompositions are still in use, and if one is deleted
having it removed from the list of valid pio decomps.

Should address #1621

### CODE REVIEW

**Comment by AaronDonahue** [comment]
@ndkeen , this hasn't been reviewed yet, but you are welcome to grab the branch and merge it to your to see how well it addresses the memory usage you are seeing at ne120.

**Comment by PeterCaldwell** [comment]
In addition to reducing our overall memory footprint, I'm also curious whether this fixes our memory leak when writing output!

**Comment by AaronDonahue** [comment]
@PeterCaldwell , unfortunately, my preliminary checks seem to indicate the output mem-leak is still an issue.  I'm investigating that now.

**Comment by ndkeen** [comment]
I added this to my branch. It does look like ne30 now using less memory (can run with spa on 1 node). I do see the slight memory growth per snapshot.

But then for ne120, I'm getting some fails.  I assume you haven't tried with ne120.

```
1536: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
1536: 
1536: Backtrace for this error:
1536: #0  0x14a8e35c5d6f in ???
1536: #1  0x14a8e6e5b911 in ???
1536: #2  0x14a8e6e33b77 in ???
1536: #3  0x14a8e58e3398 in ???
1536: #4  0x14a8e5919eb5 in ???
1536: #5  0x14a8e591ac75 in ???
1536: #6  0xf040d6 in pio_swapm
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/externals/scorpio/src/clib/pio_spmd.c:404
1536: #7  0xed0e05 in compute_counts
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/externals/scorpio/src/clib/pio_rearrange.c:824
1536: #8  0xed2537 in box_rearrange_create
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/externals/scorpio/src/clib/pio_rearrange.c:1596
1536: #9  0xecbfa8 in PIOc_InitDecomp
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/externals/scorpio/src/clib/pioc.c:771
1536: #10  0xe8756a in pio_initdecomp_internal
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/externals/scorpio/src/flib/piolib_mod.F90:600
1536: #11  0xe8756a in __piolib_mod_MOD_pio_initdecomp_dof_i8
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/externals/scorpio/src/flib/piolib_mod.F90:625
1536: #12  0xe879d5 in __piolib_mod_MOD_pio_initdecomp_dof_i4
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/externals/scorpio/src/flib/piolib_mod.F90:535
1536: #13  0xfdd8ac in __scream_scorpio_interface_MOD_get_decomp
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/components/scream/src/share/io/scream_scorpio_interface.F90:977
1536: #14  0xfde86e in __scream_scorpio_interface_MOD_set_decomp
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/components/scream/src/share/io/scream_scorpio_interface.F90:1055
1536: #15  0xfe71f0 in set_decomp_c2f
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/components/scream/src/share/io/scream_scorpio_interface_iso_c2f.F90:68
1536: #16  0xfe5515 in _ZN6scream7scorpio10set_decompERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/components/scream/src/share/io/scream_scorpio_interface.cpp:59
1536: #17  0x172a266 in _ZN6scream3spa12SPAFunctionsIdN6Kokkos6DeviceINS2_6OpenMPENS2_9HostSpaceEEEE27get_remap_weights_from_fileERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEiiRKNS2_4ViewIPiJNS2_11LayoutRightES6_NS2_12MemoryTraitsILj0EEEEEERNS7_14SPAHorizInterpE
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/components/scream/src/physics/spa/spa_functions_impl.hpp:392
1536: #18  0x171aba5 in _ZN6scream3SPA15initialize_implENS_7RunTypeE
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/components/scream/src/physics/spa/atmosphere_prescribed_aerosol.cpp:170
1536: #19  0x174d688 in _ZN6scream17AtmosphereProcess10initializeERKNS_4util9TimeStampENS_7RunTypeE
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/components/scream/src/share/atm_process/atmosphere_process.cpp:44
1536: #20  0x1751860 in _ZN6scream22AtmosphereProcessGroup15initialize_implENS_7RunTypeE
1536:   at /global/cfs/cdirs/e3sm/ndk/se03-may6/components/scream/src/share/atm_process/atmosphere_process_group.cpp:186
```

**Comment by AaronDonahue** [comment]
@ndkeen , thanks for trying the branch out and providing that stack trace.  I'm going to try to figure out what could be going wrong.  To answer your question, I haven't tried ne120 myself yet.

**Comment by AaronDonahue** [comment]
@ndkeen , can you point me to the branch you have merge this branch into?  When I follow the stack trace it takes me to lines of code that don't make sense given the stack trace.  Is it possible that your local branch also changed the file `scream_scorpio_interface.F90` ?  For example the stacktrace points to line `scream_scorpio_interface.F90:1055` which should point here: https://github.com/E3SM-Project/scream/pull/1643/files#diff-30eeaf24856c1191c8f6ac7ba71d316bd1f65481d13232837518d6db6805f47fR1055, but that is not calling a separate routine.  Furthermore, the next line up in the stack trace: `scream_scorpio_interface.F90:977` points to the line: https://github.com/E3SM-Project/scream/pull/1643/files#diff-30eeaf24856c1191c8f6ac7ba71d316bd1f65481d13232837518d6db6805f47fR977 which is just the end of an `if` statement.

**Comment by ndkeen** [comment]
As I noted, I've applied your changes on top of others of mine (unrelated).  Perhaps it would be easier if I made a fresh clone and tried again.

**Comment by AaronDonahue** [comment]
This latest commit should address the issue #1565 .

**Comment by jgfouca** [comment]
@AaronDonahue , that looks like the kind of error you get when a developer pushes an update to a submodule that was never published in the submodule repo. Since this PR doesn't appear to touch the EKAT submodule, I don't see why we'd get this error unless something bad is on master too. It's possible a network issue resulted in a misleading Jenkins error. Let's let the AT take another crack at it to be sure.

**Comment by AaronDonahue** [comment]
@jgfouca , this could also be my fault.  I recently force-pushed a PR that had EKAT changes.  Although I was under the impression that PR passed manual testing, so presumably it shouldn't have been a problem.  I labeled RETEST, we can see what happens after that.

**Comment by bartgol** [comment]
@AaronDonahue that error was my fault. I forgot to update the dyn grid IO pr, which was using EKAT from a sha that was not in the repo. #1644 should fix master, but you will have to rebase on master once more, to get the correct submodule SHA.

**Comment by AaronDonahue** [comment]
@bartgol , ok great, thanks for fixing that.

**Comment by AaronDonahue** [comment]
image

**Comment by AaronDonahue** [comment]
We no longer see any big memory leak for 1 file - all snapshots.  Although there is still some slight leak when we have 1 snapshot per file.  There must be something that isn't being deallocated properly that isn't covered in this PR.

**Comment by AaronDonahue** [comment]
@bartgol , all good points/questions.
1. I'll answer the last bullet first - This is using STATM, so I agree that absolute numbers are probably not helpful, just trends.  Also note, I'm using a different number of MPI ranks, the CIME runs use 96 ranks, the unit test runs use 12 I believe.  That will impact absolute numbers too.
2. The difference between blue and purple is that blue is a CIME run.  That might point to something particular with CIME runs.  I'm going to rerun with no output to see if there is an unrelated trend in the run in general.
3. x-axis is simulation hrs, not wall clock time.  The first snapshot is after initialization and the y-axis has the memory usage after the first RUN step subtracted from all values.  So this is growth from step 1.
4. As per "flattening out", yes, I wonder if I let the other simulations run longer if they too would flatten out.  It is worth noting, there is still some growth over time, even for the orange line, but it is very small.  Near the end we are talking about only a jump of 5-10KB, so not much at all.  In the beginning it is more like O(100KB).

**Comment by AaronDonahue** [comment]
image

Here are the same tests but without output.  Running with CIME shows a couple of noticeable bumps before flattening out.  I should also note, the unit test is an ne2 test with remapping ne4->ne2 SPA data, while the CIME test is ne4 with mapping ne30->ne4 SPA data.

**Comment by AaronDonahue** [comment]
@jgfouca or @bartgol , can one of you share with me the FAIL stacktrace for the AT on mappy?  I've tested this branch on quartz, which is an intel build, and it passes so I'm curious to what is causing the fail on mappy.

**Comment by bartgol** [comment]
It appears to be a double free error during `eam_pio_finalize`, at line 892:
```
      deallocate(prev_file_ptr%pio_file)
```

**Comment by AaronDonahue** [comment]
strange that didn't pop in my testing on quartz.  But that is the same issue that @ndkeen is seeing with ne120.

**Comment by ndkeen** [comment]
I tried this on cori-knl with ne120.  It failed with error below. I would verify nospa works on this machine, but just to give you potential feedback quickly:

```
 334: forrtl: error (76): Abort trap signal
 334: Image              PC                Routine            Line        Source             
 334: e3sm.exe           0000000003F1F89B  Unknown               Unknown  Unknown
 334: e3sm.exe           0000000002885120  Unknown               Unknown  Unknown
 334: e3sm.exe           0000000004087280  Unknown               Unknown  Unknown
 334: e3sm.exe           0000000004503961  Unknown               Unknown  Unknown
 334: e3sm.exe           0000000003E31395  _ZN6Kokkos4Impl10          65  Kokkos_Error.cpp
 334: e3sm.exe           0000000003B19833  _ZZN6scream4shoc9         209  Kokkos_Error.hpp
 334: e3sm.exe           0000000003B1CDEE  _ZN6scream4shoc9F         842  Kokkos_HostThreadTeam.hpp
 334: e3sm.exe           0000000003B06988  _ZN6scream4shoc9F         228  shoc_main_impl.hpp
 334: e3sm.exe           0000000003AF3280  _ZZN6scream4shoc9         364  shoc_main_impl.hpp
 334: e3sm.exe           0000000003B04D75  _ZN6Kokkos4Impl11         867  Kokkos_OpenMP_Parallel.hpp
 334: e3sm.exe           0000000003B04813  _ZN6Kokkos12paral         940  Kokkos_OpenMP_Parallel.hpp
 334: e3sm.exe           00000000019C4883  Unknown               Unknown  Unknown
 334: e3sm.exe           000000000198B13C  Unknown               Unknown  Unknown
 334: e3sm.exe           000000000195AC28  Unknown               Unknown  Unknown
 334: e3sm.exe           0000000003B04217  _ZN6Kokkos12paral         914  Kokkos_OpenMP_Parallel.hpp
 334: e3sm.exe           0000000003B03557  _ZN6scream4shoc9F         313  shoc_main_impl.hpp
 334: e3sm.exe           0000000003AF165A  _ZN6scream16SHOCM         402  atmosphere_macrophysics.cpp
 334: e3sm.exe           0000000003C792F1  _ZN6scream17Atmos          70  atmosphere_process.cpp
 334: e3sm.exe           0000000003CABD96  _ZN6scream22Atmos         211  atmosphere_process_group.cpp
 334: e3sm.exe           0000000003C792F1  _ZN6scream17Atmos          70  atmosphere_process.cpp
 334: e3sm.exe           0000000003CABD96  _ZN6scream22Atmos         211  atmosphere_process_group.cpp
 334: e3sm.exe           0000000003C792F1  _ZN6scream17Atmos          70  atmosphere_process.cpp
 334: e3sm.exe           0000000003CABD96  _ZN6scream22Atmos         211  atmosphere_process_group.cpp
 334: e3sm.exe           0000000003C792F1  _ZN6scream17Atmos          70  atmosphere_process.cpp
 334: e3sm.exe           0000000002B3ADBB  _ZN6scream7contro         984  atmosphere_driver.cpp
 334: e3sm.exe           0000000000504132  scream_run                205  scream_cxx_f90_interface.cpp
 334: e3sm.exe           00000000004FC28F  atm_comp_mct_mp_a         220  atm_comp_mct.F90
 334: e3sm.exe           00000000004297B9  component_mod_mp_         762  component_mod.F90
 334: e3sm.exe           0000000000405114  cime_comp_mod_mp_        3100  cime_comp_mod.F90
 334: e3sm.exe           0000000000428EBC  MAIN__                    208  cime_driver.F90
```

**Comment by AaronDonahue** [comment]
thats strange.  Nothing about this PR should impact SHOC, that error is in the `shoc_assumed_pdf` routine.

**Comment by ndkeen** [comment]
Yea I get same error now without spa. So maybe that's good new -- ie it's working on cori-knl with spa.  Trying again on cori with GNU.

**Comment by AaronDonahue** [comment]
@ndkeen , it would be good to know if this is an error with this PR or with `master` in general.

**Comment by ndkeen** [comment]
With GNU and spa, it also fails after 12 steps (not same error -- actually no useful error) on cori-knl. But actually, this is where it was stopping without spa anyway.  So I think this PR would be good to have. I also plotted the memory use with SPA and I think it looks good.

top_usedmem j59147518 n128

File: components/scream/src/share/atm_process/atmosphere_process_group.cpp [hunk 1]
```diff
@@ -209,6 +215,12 @@ void AtmosphereProcessGroup::run_sequential (const Real dt) {
     atm_proc->set_update_time_stamps(do_update);
     // Run the process
     atm_proc->run(dt);
+#ifdef SCREAM_HAS_MEMORY_USAGE
+    long long my_mem_usage = get_mem_usage(MB);
+    long long max_mem_usage;
+    m_comm.all_reduce(&my_mem_usage,&max_mem_usage,1,MPI_MAX);
+    m_atm_logger->info("[EAMxx::run_sequential::"+atm_proc->name()+"] memory usage: " + std::to_string(max_mem_usage) + "MB");
```
Comment (bartgol) [inline components/scream/src/share/atm_process/atmosphere_process_group.cpp:222]: Imho, reporting mem usage of all atm procs during each timestep is more of a "debug" log level than an "info" log level.

File: components/scream/src/control/atmosphere_driver.cpp [hunk 2]
```diff
@@ -494,6 +494,12 @@ initialize_fields (const util::TimeStamp& run_t0, const util::TimeStamp& case_t0
     }
   }
 
+#ifdef SCREAM_HAS_MEMORY_USAGE
+  long long my_mem_usage = get_mem_usage(MB);
+  long long max_mem_usage;
+  m_atm_comm.all_reduce(&my_mem_usage,&max_mem_usage,1,MPI_MAX);
+  m_atm_logger->info("[EAMxx::init::initialize_fields] memory usage: " + std::to_string(max_mem_usage) + "MB");
```
Comment (bartgol) [inline components/scream/src/control/atmosphere_driver.cpp:501]: These new calls to the memory usage in the middle of init sequence seems more of a "debug" log level than "info"...

File: components/scream/src/share/atm_process/atmosphere_process_group.cpp [hunk 3]
```diff
@@ -184,6 +184,12 @@ gather_internal_fields  () {
 void AtmosphereProcessGroup::initialize_impl (const RunType run_type) {
   for (auto& atm_proc : m_atm_processes) {
     atm_proc->initialize(timestamp(),run_type);
+#ifdef SCREAM_HAS_MEMORY_USAGE
+    long long my_mem_usage = get_mem_usage(MB);
+    long long max_mem_usage;
+    m_comm.all_reduce(&my_mem_usage,&max_mem_usage,1,MPI_MAX);
+    m_atm_logger->info("[EAMxx::initialize::"+atm_proc->name()+"] memory usage: " + std::to_string(max_mem_usage) + "MB");
```
Comment (bartgol) [inline components/scream/src/share/atm_process/atmosphere_process_group.cpp:191]: Here too. The AD already reports a total ATM memory usage right after all atm procs are inited. This is more fine grained, which probably is better suited for "debug" level.

File: components/scream/src/share/atm_process/atmosphere_process_group.cpp [hunk 4]
```diff
@@ -219,6 +231,12 @@ void AtmosphereProcessGroup::run_parallel (const Real /* dt */) {
 void AtmosphereProcessGroup::finalize_impl (/* what inputs? */) {
   for (auto atm_proc : m_atm_processes) {
     atm_proc->finalize(/* what inputs? */);
+#ifdef SCREAM_HAS_MEMORY_USAGE
+    long long my_mem_usage = get_mem_usage(MB);
+    long long max_mem_usage;
+    m_comm.all_reduce(&my_mem_usage,&max_mem_usage,1,MPI_MAX);
+    m_atm_logger->info("[EAMxx::finalize::"+atm_proc->name()+"] memory usage: " + std::to_string(max_mem_usage) + "MB");
```
Comment (bartgol) [inline components/scream/src/share/atm_process/atmosphere_process_group.cpp:238]: Here too: "debug" level seems better suited for fine grain stuff like this.

File: components/scream/src/control/atmosphere_driver.cpp [hunk 5]
```diff
@@ -1026,6 +1032,13 @@ void AtmosphereDriver::finalize ( /* inputs? */ ) {
 
   m_atm_logger->info("[EAMXX] Finalize ... done!");
 
+#ifdef SCREAM_HAS_MEMORY_USAGE
+  long long my_mem_usage = get_mem_usage(MB);
+  long long max_mem_usage;
+  m_atm_comm.all_reduce(&my_mem_usage,&max_mem_usage,1,MPI_MAX);
+  m_atm_logger->info("[EAMxx::finalize] memory usage: " + std::to_string(max_mem_usage) + "MB");
```
Comment (bartgol) [inline components/scream/src/control/atmosphere_driver.cpp:1039]: Debug here too?

**Review by bartgol — CHANGES_REQUESTED** [review]
I think we should scale down the log level for these fine grain details. I suggested a "debug" level.

**Comment by AaronDonahue** [comment]
@bartgol , good points on the "info" vs. "debug" level.  I made the changes and got them all, I think.  When you get a chance go ahead and re-review.

**Review by bartgol — APPROVED** [review]

**Pull request merged**